### PR TITLE
Add a new exporter for ApiGear solution

### PR DESCRIPTION
**Pull request by kamil-machnicki-siili**
Siili Auto collaborative changes, contributors:

- Aleksander Wawrzyniak
- Dorota Phan
- Kamil Machnicki

Export the VSS project to a folder containing a solution and module files used as input for ApiGear generator.

Right now, it generates code using the following templates:

- Unreal - `apigear-io/template-unreal`
- Cpp14 - `apigear-io/template-cpp14`
- Qt5 - `apigear-io/template-qt5`
- Qt6 - `apigear-io/template-qtcpp`

### CODE REVIEW

**Comment by erikbosch** [comment]
MoM:

Please review

File: docs/vspec.md
Comment (erikbosch) [inline docs/vspec.md:389]: This should better be added in a separate file like `apigear.md` in this directory. In the PR #382 a list referencing to other docs will be added, and when that one is merged this part can be moved to a new file.

**Comment by erikbosch** [comment]
Thank you for contribution. I noticed it just before the weekly VSS meeting yesterday and there I asked people to review it. Would you be interested in providing some more background on https://apigear.io/ and how the extension in this PR is supposed to be used. The COVESA VSS-project has a weekly meeting at Tuesdays 16.00 Central European Time, connection details available in the COVESA calendar at https://wiki.covesa.global/display/WIK4/COVESA+Common+Meeting+Schedule and it would be great if you would have the possibility to join and give some background to your work and answer questions, if any. If you intend to join a meeting in the near future me or one of the other VSS maintainers (@SebastianSchildt @adobekan ) could reserve some time for discussion. If you do not have the possibility to join it would be good to add some background information as a comment in this PR.

I took a quick look at the PR and I see in general no major problems. As you are first time contributor a maintainer need to approve any workflow run. In general our workflow for contributions is that we take merge decisions at the weekly VSS meetings, and aim to give members at least a week for review. As this PR introduces a new tool we may want to give a longer review period. there is a major rework of vss-tools going on in #382, so this one will need to be rebased when that one is merged, which possibly will happen next week.

Before accepting the PR I would like to know if you have the intention to continue supporting the new tool, like assisting if anyone has an issue with the tool or detects a bug, or if the tool needs to be refactored due to some general change in vss-tools. Just to make sure that the tool does not get abandoned in the near future.

File: src/vss_tools/vspec/vssexporters/vss2apigear.py
Comment (sschleemilch) [inline src/vss_tools/vspec/vssexporters/vss2apigear.py:486]: Please use `pathlib` functions. Here it would be `output.exists()`.

File: src/vss_tools/vspec/vssexporters/vss2apigear.py
Comment (sschleemilch) [inline src/vss_tools/vspec/vssexporters/vss2apigear.py:487]: Output is not a directory but a file

File: src/vss_tools/vspec/vssexporters/vss2apigear.py
Comment (sschleemilch) [inline src/vss_tools/vspec/vssexporters/vss2apigear.py:489]: pathlib `output.parent.mkdirs(exist_ok=True, parents=True)`

File: src/vss_tools/vspec/vssexporters/vss2apigear.py
Comment (sschleemilch) [inline src/vss_tools/vspec/vssexporters/vss2apigear.py:426]: If it is a file then have a look at `cli_options.py` how to specify that to make it a `pathlib` `Path` and checking whether it exists and can be read. `is_flag=false` is not required. `required=False` is also the default and can be skipped

File: src/vss_tools/vspec/vssexporters/vss2apigear.py
Comment (sschleemilch) [inline src/vss_tools/vspec/vssexporters/vss2apigear.py:154]: please add typing hints for return values aswell

**Review by sschleemilch — CHANGES_REQUESTED** [review]

File: docs/vspec.md
Comment (kamil-machnicki-siili) [reply docs/vspec.md]: Extracted to a separate file.

File: src/vss_tools/vspec/vssexporters/vss2apigear.py
Comment (kamil-machnicki-siili) [reply src/vss_tools/vspec/vssexporters/vss2apigear.py]: Used `pathlib`.

File: src/vss_tools/vspec/vssexporters/vss2apigear.py
Comment (kamil-machnicki-siili) [reply src/vss_tools/vspec/vssexporters/vss2apigear.py:487]: Output for ApiGear is actually a directory where the module and solution files will be placed.

File: src/vss_tools/vspec/vssexporters/vss2apigear.py
Comment (kamil-machnicki-siili) [reply src/vss_tools/vspec/vssexporters/vss2apigear.py]: Used `pathlib`.

File: src/vss_tools/vspec/vssexporters/vss2apigear.py
Comment (kamil-machnicki-siili) [reply src/vss_tools/vspec/vssexporters/vss2apigear.py]: Added typing hint sfor return value.

File: src/vss_tools/vspec/vssexporters/vss2apigear.py
Comment (kamil-machnicki-siili) [reply src/vss_tools/vspec/vssexporters/vss2apigear.py:426]: Paths from arguments are passed to the ApiGar which is responsible for their creation and ownership, the exporter itself doesn't actually rely on those.

**Comment by kamil-machnicki-siili** [comment]
@erikbosch thanks for your insights, for sure the exporter is meant to be used and maintained in the future. We've already had a chat within our team about the weekly VSS meeting and certainly someone from us will be there next Tuesday to give you more details on the topic and discuss the exporter and ApiGear usage.

File: src/vss_tools/vspec/vssexporters/vss2apigear.py
Comment (sschleemilch) [reply src/vss_tools/vspec/vssexporters/vss2apigear.py:487]: Well then you need a different argument. I think we have an --output-dir argument already in some exporter (protobuf?). So you could use that one or put it in cli_options and reuse it in both places

File: src/vss_tools/vspec/vssexporters/vss2apigear.py
Comment (sschleemilch) [inline src/vss_tools/vspec/vssexporters/vss2apigear.py:203]: missing types

File: src/vss_tools/vspec/vssexporters/vss2apigear.py
Comment (sschleemilch) [inline src/vss_tools/vspec/vssexporters/vss2apigear.py:215]: probably makes sense to use the available datatype enum

File: src/vss_tools/vspec/vssexporters/vss2apigear.py
Comment (kamil-machnicki-siili) [reply src/vss_tools/vspec/vssexporters/vss2apigear.py:490]: Good call, I just wanted to keep it consistent across all of the other exporters. Protobuf actually uses the `--output` and `--types-out-dir` options differently, so I created a new required argument `--output-dir` for this exporter.

File: src/vss_tools/vspec/vssexporters/vss2apigear.py
Comment (kamil-machnicki-siili) [reply src/vss_tools/vspec/vssexporters/vss2apigear.py:433]: Changed the path options types to `Path` as you suggested, and removed unnecessary `is_flag=False` and `required=False`.

File: src/vss_tools/vspec/vssexporters/vss2apigear.py
Comment (kamil-machnicki-siili) [reply src/vss_tools/vspec/vssexporters/vss2apigear.py]: Added missing types.

File: src/vss_tools/vspec/vssexporters/vss2apigear.py
Comment (kamil-machnicki-siili) [reply src/vss_tools/vspec/vssexporters/vss2apigear.py]: Changed the comparison to datatype enum.

**Comment by adobekan** [comment]
MoM:
@sschleemilch we need your approval for this PR.

File: docs/apigear.md
Comment (sschleemilch) [inline docs/apigear.md:40]: That include path is included by default when using spec/Vehicle.....vspec

File: src/vss_tools/vspec/vssexporters/vss2apigear.py
Comment (sschleemilch) [inline src/vss_tools/vspec/vssexporters/vss2apigear.py:359]: are we trying to write yaml here?
Why are we not building a correct dict and using yaml dump?
Generating yaml with raw strings looks a bit dangerous

File: src/vss_tools/vspec/vssexporters/vss2apigear.py
Comment (sschleemilch) [inline src/vss_tools/vspec/vssexporters/vss2apigear.py:416]: looks fishy to default the location to CWD .. /cppservice.

File: src/vss_tools/vspec/vssexporters/vss2apigear.py
Comment (sschleemilch) [inline src/vss_tools/vspec/vssexporters/vss2apigear.py:236]: `type` shadows a keyword. Usually that can lead to nasty bugs

File: src/vss_tools/vspec/vssexporters/vss2apigear.py
Comment (sschleemilch) [inline src/vss_tools/vspec/vssexporters/vss2apigear.py:330]: `path = directory / SOLUTION_FILENAME.format(module_name)`

File: src/vss_tools/vspec/vssexporters/vss2apigear.py
Comment (sschleemilch) [inline src/vss_tools/vspec/vssexporters/vss2apigear.py:331]: file descriptor is never closed.
Use a context manager:
```python
with open (path, "w") as f:
    f.write()
```

File: src/vss_tools/vspec/vssexporters/vss2apigear.py
Comment (sschleemilch) [inline src/vss_tools/vspec/vssexporters/vss2apigear.py:408]: `module_path = directory / module_filename`

File: src/vss_tools/vspec/vssexporters/vss2apigear.py
Comment (sschleemilch) [inline src/vss_tools/vspec/vssexporters/vss2apigear.py:428]: file.name probably does not reveal a good location. Just `file` probably does

**Comment by sschleemilch** [comment]
Let me know if you need help adapting it to the major refactoring that just got merged on master

**Comment by kamil-machnicki-siili** [comment]
> Let me know if you need help adapting it to the major refactoring that just got merged on master

Thanks @sschleemilch, I think I got it rebased properly.

File: src/vss_tools/vspec/vssexporters/vss2apigear.py
Comment (kamil-machnicki-siili) [reply src/vss_tools/vspec/vssexporters/vss2apigear.py:349]: Right, I'll rework it using the yaml module.

File: src/vss_tools/vspec/vssexporters/vss2apigear.py
Comment (sschleemilch) [inline src/vss_tools/vspec/vssexporters/vss2apigear.py:309]: We can derive whether a node has a datatype from its instance type. Pydantic ensures that a node that should have a datatype also actually has one

File: docs/apigear.md
Comment (kamil-machnicki-siili) [reply docs/apigear.md]: Removed include path.

File: src/vss_tools/vspec/vssexporters/vss2apigear.py
Comment (kamil-machnicki-siili) [reply src/vss_tools/vspec/vssexporters/vss2apigear.py]: Yeah it should rather be a working directory, changed.

File: src/vss_tools/vspec/vssexporters/vss2apigear.py
Comment (kamil-machnicki-siili) [reply src/vss_tools/vspec/vssexporters/vss2apigear.py]: Reworked it using the yaml module.

File: src/vss_tools/vspec/vssexporters/vss2apigear.py
Comment (kamil-machnicki-siili) [reply src/vss_tools/vspec/vssexporters/vss2apigear.py]: Used the function separator argument.

File: src/vss_tools/vspec/vssexporters/vss2apigear.py
Comment (kamil-machnicki-siili) [reply src/vss_tools/vspec/vssexporters/vss2apigear.py]: Changed the `type` name to `apigear_type`.

File: src/vss_tools/vspec/vssexporters/vss2apigear.py
Comment (kamil-machnicki-siili) [reply src/vss_tools/vspec/vssexporters/vss2apigear.py]: Changed.

File: src/vss_tools/vspec/vssexporters/vss2apigear.py
Comment (kamil-machnicki-siili) [reply src/vss_tools/vspec/vssexporters/vss2apigear.py]: Reworked this function completely so it no longer has a never closed file descriptor.

File: src/vss_tools/vspec/vssexporters/vss2apigear.py
Comment (kamil-machnicki-siili) [reply src/vss_tools/vspec/vssexporters/vss2apigear.py]: Changed.

File: src/vss_tools/vspec/vssexporters/vss2apigear.py
Comment (kamil-machnicki-siili) [reply src/vss_tools/vspec/vssexporters/vss2apigear.py]: Properly used the `directory` variable here.

File: src/vss_tools/vspec/vssexporters/vss2apigear.py
Comment (kamil-machnicki-siili) [reply src/vss_tools/vspec/vssexporters/vss2apigear.py]: I'm not sure if I followed you correctly, but I've removed this check and replaced it with datatype instance check: `isinstance(node.data, VSSDataDatatype)`.

**Review by sschleemilch — APPROVED** [review]
Thanks for doing all the proposed fixes. 👍 From my side it is good now 🕺

**Comment by kamil-machnicki-siili** [comment]
> Thanks for doing all the proposed fixes. 👍 From my side it is good now 🕺

Thank you @sschleemilch, I've just uploaded one last patch with changed `deprecation` handling as we had some name conflicts upon building and it should be all set from our side.

**Review by sschleemilch — APPROVED** [review]

**Comment by SebastianSchildt** [comment]
MoM 08/27 Merge

**Pull request merged**
